aws_signing_helper/pkcs11_signer.go

```diff
@@ -42,28 +75,124 @@ func contains(slice []uint, find uint) bool {
 	return false
 }
 
-// Opens a session with the PKCS #11 module
-func openPKCS11Session(lib string, slot uint) (module *pkcs11.Ctx, session pkcs11.SessionHandle, err error) {
-	var slots []uint
+// Used to enumerate slots with all token/slot info for matching
+type SlotIdInfo struct {
+	id	uint
+	info	pkcs11.SlotInfo
+	tokinfo pkcs11.TokenInfo
+}
+
+// Return true if the URI specifies the attribute and it *doesn't* match
+func mismatchAttr(uri *pkcs11uri.Pkcs11URI, attr string, val string) (res bool) {
+	var urival string
+	var ok bool
+
+	urival, ok = uri.GetPathAttribute(attr, false)
+	if ok && urival != val {
+		return true
+	}
+
+	return false
+}
+
+// Return the set of slots which match the given uri
+func matchSlots(slots []SlotIdInfo, uri *pkcs11uri.Pkcs11URI) (matches []SlotIdInfo) {
+
+	if uri == nil {
+		return slots
+	}
+
+	var urislotnr uint64
+	var urislot string
+	var ok bool
+
+	urislot, ok = uri.GetPathAttribute("slot-id", false)
+	if ok {
+		urislotnr, _ = strconv.ParseUint(urislot, 0, 32)
+	}
+
+	for _, slot := range slots {
+		if urislotnr != 0 && urislotnr != uint64(slot.id) {
+			continue
+		}
+		if mismatchAttr(uri, "token", slot.tokinfo.Label) ||
+		   mismatchAttr(uri, "model", slot.tokinfo.Model) ||
+		   mismatchAttr(uri, "manufacturer", slot.tokinfo.ManufacturerID) ||
+		   mismatchAttr(uri, "serial", slot.tokinfo.SerialNumber) ||
+   		   mismatchAttr(uri, "slot-description", slot.info.SlotDescription) ||
+   		   mismatchAttr(uri, "slot-manufacturer", slot.info.ManufacturerID) {
+			continue
+		}
+		matches = append(matches, slot)
+	}
+
+	return matches
+}
+
+// Initialize and enumerate slots in the PKCS#11 module
+func openPKCS11Module(lib string) (module *pkcs11.Ctx, slots []SlotIdInfo, err error) {
+	var slot_ids []uint
+
+	// In a properly configured system, nobody should need to override this.
+	if lib == "" {
+		lib = "p11-kit-proxy.so"
```

Comment:
Does this need to be a DLL on Windows?

Reply:
Yep, and `.dylib` on MacOS. Ideally, the underlying libraries would get this right for us automatically. We shouldn't expect every application author to do this manually for themselves. cf. https://github.com/miekg/pkcs11/issues/171